qutip/core/data/convert.pyx

```diff
@@ -259,63 +261,171 @@ cdef class _to:
-            return CSR(arg.tocsr(), shape=shape)
-        # Promote 1D lists and arguments to kets, not bras by default.
-        arr = np.array(arg, dtype=np.complex128)
-        if arr.ndim == 1:
-            arr = arr[:, np.newaxis]
-        if arr.ndim != 2:
-            raise TypeError("input has incorrect dimensions: " + str(arr.shape))
-        return csr.from_dense(dense.fast_from_numpy(arr))
+        Parameters
+        ----------
+        creators : iterable of (condition, creator, [priority])
+            An iterable of 2- or 3-tuples describing the new data layer
+            creation functions.
+            Each element can individually be a 2- or 3-tuple; they do not need
+            to be all one or the other.
 
+            Elements
+            ........
+            condition : callable (object) -> Data
+                function determining if the given object can be converted to a
+                data-layer type using this creator.
+
+            creator function: callable (object, shape) -> Data
+                The creator function. It should take an object and a shape
+                and return a data-layer type instance. The object may be
+                any object for which the condition function returned ``True``
+                when tested.
+
+            priority : positive real, optional (1)
+                The priority associated with this creator. Higher priority
+                conditions will be tested first and the first valid creator
+                (i.e. for which ``condition(object) == True``) will handle
+                the creation.
+
+        Notes
+        -----
+            Default creators are added with the following priorities:
+
+                * Objects that are instances of data-layer types are
+                  converted using ``.copy`` with priority ``100``.
+                * Objects that are ``numpy.ndarray`` instances or for
+                  which ``scipy.sparse.isspmatrix_csr`` is ``True`` are
+                  converted using an internal CSR converter with
+                  priority ``80``.
+                * Objects for which ``scipy.sparse.issparse`` is ``True``
+                  are converted using an internal CSR converter with
+                  priority ``10``.
+        """
+        _creators = []
+        for creator in creators:
+            if len(creator) == 2:
+                condition, create = creator
+                priority = 10
+            elif len(creator) == 3:
+                condition, create, priority = creator
+            if not callable(condition):
+                raise TypeError(str(condition) + " is not a callable")
+            if not callable(create):
+                raise TypeError(str(create) + " is not a callable")
+            _creators.append((condition, create, priority))
+        self._creators = sorted(_creators,
+                                key=lambda creator: creator[2],
+                                reverse=True)
```

Comment:
This `add_creators` function seems to completely overwrite all known creators.  It can always trust that its internal list is sorted, so it should probably just insert into the list, rather than replacing it.